docs/design.md

```diff
@@ -0,0 +1,88 @@
+# Design and architecture
+
+## Design
+The `image-rs` crate is a rustified and tailored version of [containers/image](https://github.com/containers/image),
+to provide a small, simple, secure, lightweight and high performance OCI container image management library for
+[Confidential Containers](https://github.com/confidential-containers).
+
+The `image-rs` crate focuses on usage scenarios to run `Confidential Containers` on Kubernetes clusters with
+[Kata Containers](https://katacontainers.io/). So it implements the
+[CRI Image Service](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1alpha2/api.proto#L119)
+interface to be easily integrated with the K8s ecosystem. Among the container image `Build`, `Ship` and `Run` stages,
+it focuses on the `Run` stage instead of `Build` and `Ship`, that is only to provide interfaces to prepare container
+images for running in a confidential environment. It may also be enhanced to support other confidential container
+runtimes on demand.
+
+The `image-rs` crate supports both encrypted and unencrypted container images since
+[Encrypted Container Image](https://github.com/opencontainers/artifacts/pull/15)
+is mandatory for many confidential container usage scenarios. It also supports image signature verification to ensure
+integrity.
+
+For confidential containers based on enhanced hardware virtualization technologies (TDX/SEV/PEF etc), there's no easy
+way to share container images among pods/containers running on the same physical node. But it's time consuming to
+download, verify, decrypt and extract container images for each pod/container. So the `image-rs` crate will explore
+technologies and innovations to reduce image preparation time. Data decryption and decompression acceleration libraries
+may be used to speed up image decryption and decompression. Image data on demand loading may also be adopted to reduce
+container startup time.
+
+### Goals
+ * Support pulling OCI/docker v2 images from remote registries compatible with the OCI distribution spec.
+ * Support downloading OCI images from local [OCI layout compatiable](https://github.com/opencontainers/image-spec/blob/main/image-layout.md) storage
+ * Support OCI image decryption
+ * Support OCI image signature verification
+ * Support seamlessly integration with Kata Agent and Attestation Agent
+ * Explore image decryption and decompression acceleration technologies
+ * Explore image on demand loading technologies
+
+
+### Non-Goals
+ * Image push, content discovery/management operations for remote registries
+ * Image encryption and signing
```

Comment:
I agree that everything involving provisioning an image is a non-goal that would bloat `image-rs` but let's make sure that we provide good documentation about how a user can setup images using other tools. `image-rs` isn't very useful if people can't easily upload encrypted/signed images.

Reply:
So maybe a "Howto" doc helps here:)

Reply:
ooi, is it not in scope as there are already existing tools to encrypt+sign or is that we don't want to bloat the binary? If it's the latter, we can simply build two binaries from the same source (one to encrypt/sign the other to decrypt/check) to avoid any packaging bloat.

If this project supported encrypting/signing, that would make writing unit tests easier. But at an integration level, the CI should encrypt with other tooling too! ;)

Reply:
This project will leverage [ocicrypt-rs](https://github.com/containers/ocicrypt-rs) to do decryption and image signing verification, and we may focuse integration test for compatiable with existing container build tools like [skopeo](https://github.com/containers/skopeo)

Reply:
> 

We hope to focus on the most valuable features at the first stage:)
Let's get something usable then we could extend it on demand:)

Reply:
I agree with @arronwy and @jiangliu here, image distribution and building should be kept out of scope for now and that should be very clear.

 To make it even clearer, we could indeed describe how to do those steps with *other* tools and libraries.

Reply:
Yes, we added the document for generate test data in another PR which will describe how to do those steps, and will link them when related PR are merged.